### PR TITLE
Retry 400s during dataset teardown.

### LINKED ISSUE #2318: Dataset exists (409) / dataset still in use (DELETE --> 400)
`python2`:

- Still in use: `precise-truck-742:system_tests_160041019_1473898532`
- Exists: `precise-truck-742:system_tests_160041019_1473898532`

then `python3`:

- Still in use: `precise-truck-742:system_tests_160041019_1473898730`
- Exists: `precise-truck-742:system_tests_160041019_1473898730`

From https://travis-ci.org/GoogleCloudPlatform/google-cloud-python/builds/160041019

Two proposed fixes:

- Make sure the 400 is caught and retried for "still in use"
- Use dataset IDs unique to each test (if possible)

### PR DESCRIPTION
**Pull request by tseaver**
Closes #2318.

### CODE REVIEW

File: system_tests/bigquery.py [hunk 1]
```diff
@@ -62,12 +62,15 @@ def setUp(self):
...
     def tearDown(self):
         from google.cloud.storage import Bucket
+        from google.cloud.exceptions import BadRequest
+        from google.cloud.exceptions import Conflict
+        retry_400 = RetryErrors(BadRequest)
         for doomed in self.to_delete:
             if isinstance(doomed, Bucket):
-                retry = RetryErrors(Conflict)
-                retry(doomed.delete)(force=True)
+                retry_409 = RetryErrors(Conflict)
```
Comment (dhermes) [inline system_tests/bigquery.py]: Move this up with the `retry_400`?

File: system_tests/bigquery.py [hunk 2]
```diff
@@ -62,12 +62,15 @@ def setUp(self):
...
+                retry_409(doomed.delete)(force=True)
             else:
-                doomed.delete()
+                retry_400(doomed.delete)()
```
Comment (dhermes) [inline system_tests/bigquery.py]: Worth having 3 branches, one with an `isinstance(doomed, Dataset)`?

**Comment by dhermes** [comment]
@tseaver Were you able to reproduce this locally?

**Comment by tseaver** [comment]
> Were you able to reproduce this locally?

Not except by borking the delete of tables in the dataset.

File: system_tests/bigquery.py (hunk 1 above)
Comment (tseaver) [inline system_tests/bigquery.py]: Done

File: system_tests/bigquery.py (hunk 2 above)
Comment (tseaver) [inline system_tests/bigquery.py]: Done

**Comment by dhermes** [comment]
Gotcher. Thanks for the tip. Can you add an error predicate, e.g. `exc._errors[0]['reason'] == 'resourceInUse'`

```python
>>> try:
...     dataset.delete()
... except Exception as exc:
...     pass
...
>>> import json
>>> print(json.dumps(exc.__dict__, indent=2))
{
  "message": "Dataset {PROJ}:create_table{TESTID} is still in use (DELETE https://www.googleapis.com/bigquery/v2/projects/{PROJ}/datasets/create_table{TESTID})",
  "_errors": [
    {
      "domain": "global",
      "message": "Dataset {PROJ}:create_table{TESTID} is still in use",
      "reason": "resourceInUse"
    }
  ]
}
```

**Pull request merged**